test/test_image.py

```diff
@@ -54,176 +53,202 @@ def normalize_dimensions(img_pil):
-            assert_equal(data, expected)
-            os.unlink(fpath)
-
-        with self.assertRaisesRegex(
-                RuntimeError, "No such file or directory: 'tst'"):
-            read_file('tst')
-
-    def test_read_file_non_ascii(self):
-        with get_tmp_dir() as d:
-            fname, content = '日本語(Japanese).bin', b'TorchVision\211\n'
-            fpath = os.path.join(d, fname)
-            with open(fpath, 'wb') as f:
-                f.write(content)
-
-            data = read_file(fpath)
-            expected = torch.tensor(list(content), dtype=torch.uint8)
-            assert_equal(data, expected)
-            os.unlink(fpath)
-
-    def test_write_file(self):
-        with get_tmp_dir() as d:
-            fname, content = 'test1.bin', b'TorchVision\211\n'
-            fpath = os.path.join(d, fname)
-            content_tensor = torch.tensor(list(content), dtype=torch.uint8)
-            write_file(fpath, content_tensor)
-
-            with open(fpath, 'rb') as f:
-                saved_content = f.read()
-            self.assertEqual(content, saved_content)
-            os.unlink(fpath)
-
-    def test_write_file_non_ascii(self):
-        with get_tmp_dir() as d:
-            fname, content = '日本語(Japanese).bin', b'TorchVision\211\n'
-            fpath = os.path.join(d, fname)
-            content_tensor = torch.tensor(list(content), dtype=torch.uint8)
-            write_file(fpath, content_tensor)
-
-            with open(fpath, 'rb') as f:
-                saved_content = f.read()
-            self.assertEqual(content, saved_content)
-            os.unlink(fpath)
+@pytest.mark.parametrize('img_path', [
+    pytest.param(jpeg_path, id=_get_safe_image_name(jpeg_path))
+    for jpeg_path in get_images(IMAGE_ROOT, ".jpg")
+])
+@pytest.mark.parametrize('conversion', [
+    (None, ImageReadMode.UNCHANGED),
+    ("L", ImageReadMode.GRAY),
+    ("RGB", ImageReadMode.RGB),
+])
+def test_decode_jpeg(img_path, conversion):
+    pil_mode, mode = conversion
+
+    with Image.open(img_path) as img:
+        is_cmyk = img.mode == "CMYK"
+        if pil_mode is not None:
+            if is_cmyk:
+                # libjpeg does not support the conversion
+                pytest.xfail("Decoding a CMYK jpeg isn't supported")
+            img = img.convert(pil_mode)
+        img_pil = torch.from_numpy(np.array(img))
+        if is_cmyk:
+            # flip the colors to match libjpeg
+            img_pil = 255 - img_pil
+
+    img_pil = normalize_dimensions(img_pil)
+    data = read_file(img_path)
+    img_ljpeg = decode_image(data, mode=mode)
+
+    # Permit a small variation on pixel values to account for implementation
+    # differences between Pillow and LibJPEG.
+    abs_mean_diff = (img_ljpeg.type(torch.float32) - img_pil).abs().mean().item()
+    assert abs_mean_diff < 2
+
+    with pytest.raises(RuntimeError, match="Expected a non empty 1-dimensional tensor"):
```

Comment:
the assertions will be run on all parametrized input, which is a bit  overkill. We can extract them out in a new test function, like what was done for `test_encode_jpeg_errors`

Reply:
@NicolasHug , done as suggested!